TERRAFORM.md

```diff
@@ -0,0 +1,96 @@
+# Terraform Best Practices and Guidelines
+
+General terraform guidelines and best practices
+
+## Table of Contents
+
+- [Table of Contents](#table-of-contents)
+- [Introduction](#introduction)
+- [Code Organization](#code-organization)
+  - [Reusable Module Usage](#reusable-module-usage)
+- [State Management](#state-management)
+  - [Remote State](#remote-state)
+    - [State key](#state-key)
+  - [State Locking](#state-locking)
+  - [State migration](#state-migration)
+  - [Module Guidelines](#module-guidelines)
+- [Code formatting](#code-formatting)
+
+## Introduction
+
+Terraform is a powerful tool for provisioning and managing infrastructure as code. To ensure maintainability, scalability, and security, it's important to follow best practices and guidelines when working with Terraform.
+
+## Code Organization
+
+The high level code organization should look like this
+
+```text
+|- modules/                   # terraform reusable modules for the whole repository
+|- ORGANIZATIONS_OR_ACCOUNT   # e.g. AWS Account name
+|           |- modules        # terraform reusable modules for this organization/account
+|           |- shared         # shared resources
+|           |- STAGING        # resource pro staging
+```
+
+Then it should mirror the hierarchical structure of the infrastructure. Note that each terraform module (project) should be kept small.
+
+### Reusable Module Usage
+
+Use modules to encapsulate and reuse configurations. This promotes DRY (Don't Repeat Yourself) principles and makes your configurations more manageable.
+
+- Create modules for common resources.
+- Use versioned modules from the Terraform Registry when possible.
+- Use modules to share global variales between terraform modules
+
+## State Management
+
+### Remote State
+
+We use remote state management using AWS S3 backend
+
+- Encrypt the state file at rest (see [Terraform Documentation](https://developer.hashicorp.com/terraform/language/settings/backends/s3#encrypt)).
+- Put the state backend definition in `terraform.tf` file
+
+#### State key
+
+We use the terraform module definition file path as state key in S3 prefixed with a version string `v1`; e.g. `v1/infra-terraform-bgdi/systems-prod/map.terraform.tf`
+
+The version prefix is to avoid any risk of key collision in the future in case of code structure re-organization.
```

Comment:
Can we enforce correct key usage with the CI?

Reply:
Same I would not try to add a complex CI to enforce everything